hazelcast/src/main/java/com/hazelcast/sql/impl/expression/predicate/ComparisonPredicate.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.predicate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpression;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Implements evaluation of SQL comparison predicates.
+ *
+ * @see ComparisonMode
+ */
+public final class ComparisonPredicate extends BiExpression<Boolean> implements IdentifiedDataSerializable {
+
+    private ComparisonMode mode;
+
+    public ComparisonPredicate() {
+        // No-op.
+    }
+
+    private ComparisonPredicate(Expression<?> left, Expression<?> right, ComparisonMode mode) {
+        super(left, right);
+        this.mode = mode;
+    }
+
+    public static ComparisonPredicate create(Expression<?> left, Expression<?> right, ComparisonMode comparisonMode) {
+        assert left.getType().equals(right.getType());
+        return new ComparisonPredicate(left, right, comparisonMode);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_COMPARISON;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @SuppressFBWarnings(value = "NP_BOOLEAN_RETURN_NULL", justification = "Any SQL expression may return null")
+    @Override
+    public Boolean eval(Row row, ExpressionEvalContext context) {
+        Object left = operand1.eval(row, context);
+        if (left == null) {
+            return null;
+        }
+
+        Object right = operand2.eval(row, context);
+        if (right == null) {
+            return null;
+        }
+
+        Comparable leftComparable = (Comparable) left;
```

Comment:
It seems that we do not support comparison for `ANY` type. What is the reason for this? If two objects have the same type and are comparable, then we can compare them, aren't we?

Reply:
Initially I started working on it, but there were too many open questions. For instance, `OBJECT = OBJECT` basically means we are comparing one object of unknown type to another object of unknown type, we can't be 100% sure they are comparable during query validation. Next, if we would allow comparisons for objects of unknown types, it's also logical to allow comparisons of a known type to an unknown one, like 'VARCHAR = OBJECT'. It's not clear should we use `equals` instead of `compareTo`, if one or both of the operands are not implementing `Comparable` and how that all would interact with indexes. So I decided to postpone it, if you think it's critical to have it in 4.1, I may continue the work.

Reply:
I think we should create an issue, and try squeezing it into 4.1 once other more important expressions are ready. If not, we can do that in 4.2. But we need a ticket not to forget about this limitation.

I am not fully sure about non-equality comparisons, but not allowing comparing objects doesn't seem good for the engine.

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17302 for 4.2